### PR TITLE
Initial prototype of Activator.CreateFactory

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
__\*\* DRAFT \*\* DRAFT \*\* DRAFT \*\*__

Contributes to https://github.com/dotnet/runtime/issues/36194. Doesn't fully resolve that issue since there's still a second set APIs which aren't present in this PR.

This is the follow-up to https://github.com/dotnet/runtime/pull/32520, which takes advantage of that refactoring to introduce new `Activator.CreateFactory(Type)` and `Activator.CreateFactory<T>` APIs. Think of a factory as a cached version of `CreateInstance`, where we've already done all the initial work and can keep invoking our factory over and over every time we need a new instance.

As a reminder, the purpose of this new API is two-fold. First, it allows callers to remove their own usage of ref-emit. Second, it tries to avoid spinning up the JIT as much as possible. `Activator.CreateFactory(Type)` should be JIT-free in nearly all cases, including for value types. `Activator.CreateFactory<T>` may spin up the JIT for value types _T_.

I've also hooked this factory up through _System.Text.Json_ to show how it can replace their existing usage of `DynamicMethod` in some scenarios.

__This is a draft PR and isn't fully ready for review.__ There's still some code cleanup that needs to take place, but it should give a good indication of how things should look when they're fully finished.

To-do:

- [ ] Make mono runtime compile and tests pass
- [ ] Figure out whether the `Delegate` changes were appropriate or if they should be backed out
- [ ] Fix layering in _Activator.cs_ and _Activator.CoreCLR.cs_
- [ ] Fix suppressions and dynamic dependency attributes introduced here
- [ ] Add API docs

There are some other uses of `Activator.CreateInstance` ([see here](https://github.com/dotnet/runtime/blob/cd5619bacc89deff7c57572a3d8f1349621c15d0/src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComTypeClassDesc.cs#L24) for one example), but I didn't change those call sites as part of this PR since I don't think they're on hot paths.

/cc @layomia and @Jozkee as area owners of _System.Text.Json_ so that they can see how this would look in practice. If you're wondering about the double-delegate pattern, see https://github.com/dotnet/runtime/issues/45408 and https://github.com/dotnet/runtime/issues/36194#issuecomment-736174298 for context.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs [hunk 1]
```diff
@@ -27,6 +31,9 @@ internal sealed class ReflectionEmitMemberAccessor : MemberAccessor
                 return null;
             }
 
+#if NET6_0 // should really be NET6_0_OR_GREATER
+            return new JsonClassInfo.ConstructorDelegate(Activator.CreateFactory(type));
```
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs:35]: There are 20 instance of ConstructorDelegate in 7 files. I think we should just replace it with `Func<object>`.

File: src/libraries/System.Runtime/tests/System/ActivatorTests.Generic.cs [hunk 2]
```diff
@@ -60,6 +76,82 @@ public partial class ActivatorTests
...
+        }
+
+        [Fact]
+        public void CreateFactoryT_OfValueTypeWithPublicDefaultCtor_CallsDefaultCtor()
+        {
+            Func<StructWithPublicDefaultConstructor> factory = Activator.CreateFactory<StructWithPublicDefaultConstructor>();
+            Assert.NotNull(factory);
+            StructWithPublicDefaultConstructor instance = factory();
+            Assert.True(instance.ConstructorInvoked);
+            Assert.True(IsAddressOnLocalStack(instance.AddressPassedToConstructor)); // using Func<T>, value type ctor is called using ref to stack local, no boxing
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/ActivatorTests.Generic.cs:124]: This is implementation detail. I can imagine some imagine some implementations may choose to allocate the box to keep things simple.

File: src/libraries/System.Private.CoreLib/src/System/Activator.RuntimeType.cs [hunk 3]
```diff
@@ -152,5 +152,50 @@ public static partial class Activator
...
+
+            ActivationFactory factory = new ActivationFactory(rt);
+            if (!nonPublic && !factory.CtorIsPublic)
+            {
+                throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, rt));
+            }
+
+            // Then create a delegate to factory.CreateInstance, closed over the "this" parameter.
+
+            return (Func<object?>)factory.CreateDelegate((RuntimeType)typeof(Func<object?>));
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Activator.RuntimeType.cs:174]: Why can't this just call `new Func<object?>(factory.CreateInstance)` ? This looks slow and complicated for no good reason.

File: src/libraries/System.Private.CoreLib/src/System/Activator.RuntimeType.cs [hunk 4]
```diff
@@ -152,5 +152,50 @@ public static partial class Activator
...
+        }
+
+        [DynamicDependency("#ctor", typeof(ActivationFactory<>))]
+        public static Func<T> CreateFactory<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]T>()
+        {
+            if (typeof(T).IsValueType)
+            {
+                IActivationFactory factory = (IActivationFactory)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(ActivationFactory<>), (RuntimeType)typeof(T));
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Activator.RuntimeType.cs:182]: CreateInstanceForAnotherGenericParameter is AOT unfriendly API that depends on JIT. Can this be just something something like this?
```
class StructFactory<T>
{
    T CreateInstance()
    {
         T value;
         _pfnConstructor(&value);
         return value;
    }
}
```
return new StructFactory<T>().;
```
?

**Comment by ViktorHofer** [comment]
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

**Comment by steveharter** [comment]
@GrabYourPitchforks based on prior discussion I assume the perf numbers are the same as ref.emit (once warmed up). Is that correct?

**Comment by GrabYourPitchforks** [comment]
@steveharter Yes, it should meet or exceed ref emit in terms of wall clock time + memory utilization.

**Pull request closed without merging**